ethree-common/src/main/java/com/virgilsecurity/android/common/storage/cloud/CloudTicketStorage.kt

```diff
@@ -0,0 +1,217 @@
+ *     this list of conditions and the following disclaimer in the documentation
+ *     and/or other materials provided with the distribution.
+ *
+ *     (3) Neither the name of virgil nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.virgilsecurity.android.common.storage.cloud
+
+import com.virgilsecurity.android.common.storage.local.KeyStorageLocal
+import com.virgilsecurity.android.common.model.Ticket
+import com.virgilsecurity.android.common.util.Const
+import com.virgilsecurity.common.model.Data
+import com.virgilsecurity.common.util.toHexString
+import com.virgilsecurity.crypto.foundation.GroupSessionMessage
+import com.virgilsecurity.keyknox.KeyknoxManager
+import com.virgilsecurity.keyknox.client.*
+import com.virgilsecurity.sdk.cards.Card
+import com.virgilsecurity.sdk.crypto.VirgilPublicKey
+import com.virgilsecurity.sdk.jwt.contract.AccessTokenProvider
+import java.net.URL
+
+/**
+ * CloudTicketStorage
+ */
+class CloudTicketStorage(
+        accessTokenProvider: AccessTokenProvider,
+        private val keyStorageLocal: KeyStorageLocal,
+        baseUrl: String? = Const.VIRGIL_BASE_URL
+) {
+
+    private val identity: String get() = keyStorageLocal.identity
+    private val keyknoxManager: KeyknoxManager
+
+    init {
+        val keyknoxClient = KeyknoxClient(accessTokenProvider, URL(baseUrl))
+        this.keyknoxManager = KeyknoxManager(keyknoxClient)
+    }
+
+    internal fun store(ticket: Ticket, cards: Collection<Card>) {
+        val selfKeyPair = keyStorageLocal.load()
+
+        val groupMessage = ticket.groupMessage
+
+        val sessionId = groupMessage.sessionId.toHexString()
+        val epoch = groupMessage.epoch
+        val ticketData = groupMessage.serialize()
+
+        val identities = cards.map { it.identity }
+        val publicKeys = cards.map { it.publicKey }
+
+        val params = KeyknoxPushParams(identities + this.identity,
+                                       GROUP_SESSION_ROOT,
+                                       sessionId,
+                                       "\\$epoch")
```

Comment:
Should be without extra slash

Reply:
done
will check why compatibility test passed successfully